### PR TITLE
SURFACESDL: Fix calculating the mouse dirty rectangle

### PR DESCRIPTION
**Pull request by ccawley2011**

### CODE REVIEW

**Comment by bluegr** [comment]
Can you provide some more information? What has been fixed here? Can you provide a use case? Could you provide some screenshots that illustrate the problem?

**Comment by ccawley2011** [comment]
Some background information can be found in this issue: https://bugs.scummvm.org/ticket/14024

This isn't a problem with SDL2 since the list of dirty rectangles is ignored by the SDL_UpdateRects replacement, but I'm not sure why it was only visible with 32-bit displays. This PR attempts to fix the issue by moving the addition of both mouse dirty rects to before the game screen or overlay is blitted.

**Review by phcoder — APPROVED** [review]

**Comment by sev-** [comment]
Thank you!

**Comment by lotharsm** [comment]
Unfortunately, it seems that this commit introduced a regression either in SURFACESDL itself or the KYRA engine: https://bugs.scummvm.org/ticket/14147

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 1]
```diff
@@ -2128,65 +2124,67 @@ void SurfaceSdlGraphicsManager::blitCursor() {
...
-	_mouseBackup.y = dst.y - hotY;
-	_mouseBackup.w = dst.w;
-	_mouseBackup.h = dst.h;
-
 	// We draw the pre-scaled cursor image, so now we need to adjust for
 	// scaling, shake position and aspect ratio correction manually.
 
-	dst.x += _currentShakeXOffset;
-	dst.y += _currentShakeYOffset;
+	dst.x = _mouseNextRect.x + _currentShakeXOffset;
```
Comment (lephilousophe) [inline backends/graphics/surfacesdl/surfacesdl-graphics.cpp:2184]: This is added twice: one in undrawMouse and now here.

File: backends/graphics/surfacesdl/surfacesdl-graphics.cpp [hunk 2]
```diff
@@ -2128,65 +2124,67 @@ void SurfaceSdlGraphicsManager::blitCursor() {
...
+	// Add the area covered by the mouse cursor to the list of dirty rects if
+	// we have to redraw the mouse, or if the cursor is alpha-blended since
+	// alpha-blended cursors will happily blend into themselves if the surface
+	// under the cursor is not reset first
+	//
+	// The mouse is undrawn using virtual coordinates, i.e. they may be
+	// scaled and aspect-ratio corrected.
+
+	if (_mouseLastRect.w != 0 && _mouseLastRect.h != 0)
+		addDirtyRect(_mouseLastRect.x - hotX, _mouseLastRect.y - hotY, _mouseLastRect.w, _mouseLastRect.h, _overlayInGUI);
```
Comment (lephilousophe) [inline backends/graphics/surfacesdl/surfacesdl-graphics.cpp:2159]: if hotX has changed between last frame and now, this calculation is invalid.
I believe it causes bug https://bugs.scummvm.org/ticket/14147

I wonder if everything should be calculated here and not in drawMouse

**Comment by moralrecordings** [comment]
@ccawley2011 could you give an example of how to reproduce the issue with a 32-bit game? The bit making internUpdateScreen() call undrawMouse() every time is causing performance issues in Director (there's a lot of ad-hoc places which can call updateScreen, relying a bit on the assumption that an empty dirty rect is a near-zero cost).

**Comment by ccawley2011** [comment]
> @ccawley2011 could you give an example of how to reproduce the issue with a 32-bit game? The bit making internUpdateScreen() call undrawMouse() every time is causing performance issues in Director (there's a lot of ad-hoc places which can call updateScreen, relying a bit on the assumption that it's near-zero cost when there's no dirty rects).

It can be replicated in the launcher when compiling on Linux with SDL 1.2 and changing the call to `SDL_SetVideoMode` to use 32bpp instead of 16bpp or 8bpp - without this PR, you should be able to notice the cursor graphic getting clipped when moving the mouse from side to side.

**Pull request merged**